framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java

```diff
@@ -1342,22 +1341,32 @@ protected List<Attribute> addJoins(StringBuilder str, Collection<JoinBuilder<Sea
                     onClause.append("?");
                     joinAttrList.add(join.getFirstAttributes()[i]);
                 } else {
-                    onClause.append(joinedTableNames.getOrDefault(join.getFirstAttributes()[i].table, join.getFirstAttributes()[i].table))
-                    .append(".")
-                    .append(join.getFirstAttributes()[i].columnName);
+                    if (join.getFirstAttributes()[i].table == null && join.getFirstAttributes()[i].value == null) {
+                        onClause.append(joinedTableNames.getOrDefault(join.getSecondAttribute()[i].table, join.getFirstAttributes()[i].table))
+                                .append(".")
+                                .append(join.getSecondAttribute()[i].columnName);
+                    } else {
+                        onClause.append(joinedTableNames.getOrDefault(join.getFirstAttributes()[i].table, join.getFirstAttributes()[i].table))
+                                .append(".")
+                                .append(join.getFirstAttributes()[i].columnName);
+                    }
                 }
-                onClause.append("=");
-                if (join.getSecondAttribute()[i].getValue() != null) {
-                    onClause.append("?");
-                    joinAttrList.add(join.getSecondAttribute()[i]);
+                if (join.getFirstAttributes()[i].table == null && join.getFirstAttributes()[i].value == null) {
```

Comment:
Need to handle the case when the `secondAttribute` is null.
```suggestion
                if ((join.getFirstAttributes()[i].table == null && join.getFirstAttributes()[i].value == null)
                        || (join.getSecondAttribute()[i].table == null && join.getSecondAttribute()[i].value == null)) {
```

Reply:
@vishesh92 is this when the placement of nicSearch.getEntity().getRemoved and userVmSearchBuilder.entity().setLong(null) is reversed in https://github.com/apache/cloudstack/pull/9547/files/f8b6e6d133fdb83ca84e098518e7b3ee87b20f97#diff-4331e50c256338af39135a4a3201b3a99d8f0c08b226e5efb9789c4866d11ceaR1470 ? I am just trying to under stand why this check is required.

Reply:
Yes

Reply:
So I had done the test by reversing the placement of the 2 args, but first and second argument still remained the same i.e., firstArgument was null.null and second was nics.removed no matter the placement.

Reply:
I used the below code to reproduce the error:
```java
            userVmSearchBuilder.join("nicSearch", nicSearch, JoinBuilder.JoinType.LEFT,
                    JoinBuilder.JoinCondition.AND,
                    nicSearch.entity().getInstanceId(), userVmSearchBuilder.entity().getId(),
                    nicSearch.entity().getRemoved(), userVmSearchBuilder.entity().setLong(null),
                    nicSearch.entity().setLong(null), userVmSearchBuilder.entity().getRemoved());
```
To add another condition during join`AND vm_instance.removed IS NULL`. But the query becomes `AND vm_instance.removed = nicSearch.null`.

Reply:
addition of the following join condition: `nicSearch.entity().setLong(null), userVmSearchBuilder.entity().getRemoved())` for testing, yeilds `nicSearch.removed IS NULL AND null.removed IS NULL INNER JOIN account accountSearch ON ` and addition of the check doesn't solve the issue. 
However, on using the following:
```
userVmSearchBuilder.join("nicSearch", nicSearch, JoinBuilder.JoinType.LEFT, JoinBuilder.JoinCondition.AND, nicSearch.entity().getInstanceId(), userVmSearchBuilder.entity().getId(), userVmSearchBuilder.entity().setLong(null),  nicSearch.entity().getRemoved());
```

I cant seem to reproduce the issue